src/platform_impl/linux/x11/mod.rs

```diff
@@ -239,15 +211,41 @@ impl<T: 'static> EventLoop<T> {
 
         xconn.update_cached_wm_info(root);
 
-        let poll = Poll::new().unwrap();
-        let waker = Arc::new(Waker::new(poll.registry(), USER_REDRAW_TOKEN).unwrap());
+        // Create an event loop.
+        let event_loop =
+            Loop::<EventLoopState<T>>::try_new().expect("Failed to initialize the event loop");
+
+        // Create the X11 event dispatcher.
+        let source = X11Source::new(xconn.x11_fd, calloop::Interest::READ, calloop::Mode::Level);
 
-        poll.registry()
-            .register(&mut SourceFd(&xconn.x11_fd), X_TOKEN, Interest::READABLE)
-            .unwrap();
+        let handle = event_loop.handle();
+        handle
+            .insert_source(source, |_, _, _| Ok(calloop::PostAction::Continue))
+            .expect("Failed to register the X11 event dispatcher");
 
-        let (user_sender, user_channel) = std::sync::mpsc::channel();
-        let (redraw_sender, redraw_channel) = std::sync::mpsc::channel();
+        let (user_sender, user_channel) = channel();
+        let (redraw_sender, redraw_channel) = channel();
+
+        // Create a dispatcher for the redraw channel such that we can dispatch it independent of the
+        // event loop.
+        let redraw_dispatcher =
+            Dispatcher::<_, EventLoopState<T>>::new(redraw_channel, |ev, _, state| {
+                if let ChanResult::Msg(window_id) = ev {
+                    state.redraw_events.push_back(window_id);
+                }
+            });
+
+        // Register the channels with the event loop.
+        handle
+            .insert_source(user_channel, |ev, _, state| {
+                if let ChanResult::Msg(user) = ev {
+                    state.user_events.push_back(user);
+                }
+            })
+            .expect("Failed to register the user event channel with the event loop");
+        handle
+            .register_dispatcher(redraw_dispatcher.clone())
+            .expect("Failed to register the redraw event channel with the event loop");
```

Comment:
Could we have source/dispatcher insert with the transfer point created, so `user_sender` -> `insert_source` -> `empty line` -> `redraw_sender` -> `dispatcher`, it's just  looks weird that we create dispatcher, then insert user source, and then insert dispatcher.